src/Translations/TranslationExtractor.php

```diff
@@ -0,0 +1,145 @@
+<?php declare(strict_types=1);
+
+namespace Becklyn\PhpCs\Translations;
+
+use Becklyn\PhpCs\Translations\Integration\NameResolverIntegration;
+use Becklyn\PhpCs\Translations\Visitor\BackendTranslatorVisitor;
+use Becklyn\PhpCs\Translations\Visitor\CallbackValidationVisitor;
+use Becklyn\PhpCs\Translations\Visitor\ClassValidationVisitor;
+use Becklyn\PhpCs\Translations\Visitor\FormOptionLabelsVisitor;
+use Becklyn\PhpCs\Translations\Visitor\PropertyValidationVisitor;
+use Symfony\Component\Finder\Finder;
+use Translation\Extractor\Extractor;
+use Translation\Extractor\FileExtractor\FileExtractor;
+use Translation\Extractor\FileExtractor\PHPFileExtractor;
+use Translation\Extractor\FileExtractor\TwigFileExtractor;
+use Translation\Extractor\Model\SourceCollection;
+use Translation\Extractor\Model\SourceLocation;
+use Translation\Extractor\Visitor\Php\Symfony\ContainerAwareTrans;
+use Translation\Extractor\Visitor\Php\Symfony\ContainerAwareTransChoice;
+use Translation\Extractor\Visitor\Php\Symfony\FlashMessage;
+use Translation\Extractor\Visitor\Php\Symfony\FormTypeChoices;
+use Twig\Environment;
+use Twig\Loader\FilesystemLoader;
+
+class TranslationExtractor
+{
+    /**
+     * @param array $dirs
+     *
+     * @return string
+     */
+    public function extract (array $dirs) : array
+    {
+        if (empty($dirs))
+        {
+            return new SourceCollection();
+        }
+
+        $extractor = new Extractor();
+        $extractor->addFileExtractor($this->createPhpExtractor());
+        $extractor->addFileExtractor($this->createTwigExtractor($dirs));
+
+        $messages = [];
+        /** @var SourceLocation $location */
+        foreach ($extractor->extract($this->createFinder($dirs)) as $location)
+        {
+            $domain = $location->getContext()["domain"] ?? "messages";
+            $messages[$domain][] = $location->getMessage();
+        }
+
+        return $this->dedupeMessages($messages);
+    }
+
+
+    /**
+     * Dedupes the messages
+     *
+     * @param array $domains
+     *
+     * @return array
+     */
+    private function dedupeMessages (array $domains): array
+    {
+        $result = [];
+
+        foreach ($domains as $group => $messages)
+        {
+            $filtered = [];
```

Comment:
This loop logic can be simplified a bit.

```suggestion
            $filtered = \array_filter($messages);
            
            if (!empty($filtered))
            {
            	$result[$group] = \array_keys($filtered);
            }
```

Reply:
This is not the same code. You are still indexing by key, whereas I am indexing by value.
So the equivalent code would look like this:

```php

$filtered = \array_keys(\array_flip(\array_filter($messages)));
if (!empty($filtered))
{
    $result[$group] = $filtered;
}
```

... which is 
a) 2 times slower as we need 3 iterations now instead of 1
b) we can discuss whether it is more readable. I would argue that it's not. 😉